generation of RXified API

phiz71 opened the pull request:
- changing FUTURE_GENERATION option to RX_INTERFACE
- creating specific templates for RX
- generating a rx sample server
- using vertx 3.4.1 in generated servers

philippkumar commented on modules/vertx-swagger-codegen/src/main/resources/javaVertXServer/apiRXVerticle.mustache:
```diff
@@ -0,0 +1,85 @@
+                    {{/allParams}}
+                {{/hasParams}}
+                service.{{operationId}}({{#hasParams}}{{#allParams}}{{paramName}}{{#hasMore}}, {{/hasMore}}{{/allParams}}{{/hasParams}}).subscribe(
+                    onSuccess -> {
```
I'd rename "onSuccess" to "result" (or something similar), as it holds the result object.
Same for "onError", which is better named "error" IMHO.

philippkumar commented on modules/vertx-swagger-codegen/src/main/java/com/github/phiz71/vertx/swagger/codegen/JavaVertXServerGenerator.java:
```diff
@@ -131,12 +131,14 @@ public String getHelp() {
+        if (additionalProperties.containsKey(RX_INTERFACE_OPTION)) {
             apiTemplateFiles.clear();
-            apiTemplateFiles.put("apiFuture.mustache", // the template to use
+            apiTemplateFiles.put("apiRX.mustache", // the template to use
```
Looks good, although I am wondering how well this approach of multiple mustache files (that might actually share several things) will scale in the future. For example, I have a feature in my fork where I generate something like `@Inject PetApi service` instead of `PetApi service = new PetApiImpl()`, resulting in a file "injectVerticle.mustache". This will make it hard to maintain in the future.

Is there a way to modularize the template files more?

I think it is fine how we have it currently, but just wanted to raise it so we can think about it.

phiz71 replied:
you can import templates into others. So I think it's possible to split the `service.subscribe/service.handler` into 2 small templates and then import them into a single api template.
Maybe it would be easier to do that, once the 2 PR are committed ?

philippkumar replied:
I agree, nothing that is necessary right now.

philippkumar commented on modules/vertx-swagger-codegen/src/main/resources/javaVertXServer/apiRXVerticle.mustache:
```diff
@@ -0,0 +1,85 @@
+                {{/returnType}}
+                    },
+                    onError -> {
+                        LOGGER.error("Error in " + {{#vendorExtensions}}{{x-serviceId-varName}}{{/vendorExtensions}}, onError);
```
I'd say we define our error handling approach and then follow it in all mustache files.

My current recommendation is to log all unexpected errors, so errors that are not of type MainApiException. I did it like this in #34. Let me know, what you think.

phiz71 replied:
I will do the same as you in your last commit in #34

philippkumar commented on sample/petstore-vertx-rx-server/src/main/java/io/swagger/server/api/MainApiVerticle.java:
```diff
@@ -0,0 +1,75 @@
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.web.Router;
+
+public class MainApiVerticle extends AbstractVerticle {
```
Not related to this PR, but I wonder if we should make the generation of MainApiVerticle optional.

In my use case, I want to do dependency injection on verticles. I do that by scanning for classes of type verticle, then instantiating them in a spring context (so dependencies can be injected) and then deploy those existing verticle instances into vert.x. That does not work well with MainApiVerticle, and I would have to exclude it when scanning.

Not sure if my way of dealing with dependency injection in vert.x is common, though.

phiz71 replied:
I can't help you with DI in Vert.X, but I do agree for the optional generation. And I would add that the fact that this MainVerticle also embed an http server can be annoying.
I will create an issue to remember.

philippkumar commented on sample/petstore-vertx-rx-server/src/main/java/io/swagger/server/api/verticle/PetApiImpl.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Override
+    public Single<Void> updatePet(Pet body) {
+        return Single.just(null);
```
In rxjava2, null is no longer an allowed item to emit in an observable stream.

That being said, I think vertx-rx is currently depending on rxjava 1.x. In this case, your code is correct, but will break at some point.

The proper way to do this in rxjava2 would be to use Completable, I think: http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Completable.html

I think Completable is also present in rxjava 1.x. If so, I recommend to use that instead of `Single<Result>` in cases where there is no result.

phiz71 commented:
@philippkumar : could you have a quick look ? 
I've replaced Single<Void> by Completable, merged what you did and extracted RX and Async specific codes into templates that I import into a common api template.

philippkumar commented:
Looks good to me.

incubus8 commented:
Hi guys, this is awesome work 👏 👏 👏. When can this be merged and released?

phiz71 commented:
the new version is to be released soon. I hope tomorrow :)
